### PR TITLE
MON-3477-new-backend-export-rrdtool

### PR DESCRIPTION
**Pull request by garnier-quentin**
* Centreon-Web Graph Display and png export is coherent  (see screenshots)
* Increase performance on server side when we get data from rrd files  (between 70% and 90%)
* Compatibility with rrdtool >= 1.5.x
* Better management of virtual metrics: you can display or not a virtual metric
* Only one color by curve (users see the same color curve)
* Remove wizard graph tour in performance view
* Fix export png for splitted graph

*old system*
screen-result-old

*new system*
screen-result-new

### CODE REVIEW

File: www/class/centreonGraphNg.class.php [hunk 1]
```diff
@@ -0,0 +1,1194 @@
...
+        $commandLine = preg_replace("/(\\\$|`)/", "", $commandLine);
+        $this->log($commandLine);
+
+        if (is_writable($this->generalOpt['debug_path']['value'])) {
+            $stderr = array('file', $this->generalOpt['debug_path']['value'] . '/rrdtool.log', 'a');
+        } else {
+            $stderr = array('pipe', 'a');
+        }
+        $descriptorspec = array(
+                            0 => array("pipe", "r"),  // stdin est un pipe processus va lire
```
Comment (sc979) [inline www/class/centreonGraphNg.class.php]:
```suggestion
                            0 => array("pipe", "r"),  // stdin - pipe process to read
```

File: www/class/centreonGraphNg.class.php [hunk 2]
```diff
@@ -0,0 +1,1194 @@
...
+                            1 => array("pipe", "w"),  // stdout est un pipe processus va ecrire
```
Comment (sc979) [inline www/class/centreonGraphNg.class.php]:
```suggestion
                            1 => array("pipe", "w"),  // stdout - pipe process to write
```

File: www/class/centreonGraphNg.class.php [hunk 3]
```diff
@@ -0,0 +1,1194 @@
...
+                            2 => $stderr // stderr est un fichier
```
Comment (sc979) [inline www/class/centreonGraphNg.class.php]:
```suggestion
                            2 => $stderr // stderr is a file
```

Comment (sc979) [reply www/class/centreonGraphNg.class.php]: please comment in english ;)

File: www/class/centreonGraphNg.class.php [hunk 4]
```diff
@@ -0,0 +1,1194 @@
...
+                        );
+
+        $process = proc_open($this->generalOpt['rrdtool_path_bin']['value'] . " - ", $descriptorspec, $pipes, null, null);
```
Comment (sc979) [inline www/class/centreonGraphNg.class.php]: seems too long

File: www/class/centreonGraphNg.class.php [hunk 5]
```diff
@@ -0,0 +1,1194 @@
...
+        $this->graphData = array(
+            'global' => $this->extraDatas,
+            'metrics' => array(),
+        );
+        foreach ($this->metrics as $metric) {
+            if ($metric['hidden'] == 1) {
```
Comment (sc979) [inline www/class/centreonGraphNg.class.php]:
```suggestion
            if ($metric['hidden'] === 1) {
```

File: www/class/centreonGraphNg.class.php [hunk 6]
```diff
@@ -0,0 +1,1194 @@
...
+                continue;
+            }
+            $this->graphData['metrics'][] = $metric;
+        }
+        foreach ($this->vmetricsOrder as $vmetricId) {
+            if ($this->vmetrics[$vmetricId]['hidden'] == 1) {
```
Comment (sc979) [inline www/class/centreonGraphNg.class.php]:
```suggestion
            if ($this->vmetrics[$vmetricId]['hidden'] === 1) {
```

File: www/class/centreonGraphNg.class.php [hunk 7]
```diff
@@ -0,0 +1,1194 @@
...
+            $rrdData = json_decode($str, true);
+        }
+        
+        $this->formatByMetrics($rrdData);
+        return $this->graphData;
+    }
+
+    /**
+     *
+     * Enter description here ...
```
Comment (sc979) [inline www/class/centreonGraphNg.class.php]: WIP

File: www/class/centreonGraphNg.class.php [hunk 8]
```diff
@@ -0,0 +1,1194 @@
...
+            } else {
+                return htmlentities($defaultValue, ENT_QUOTES, "UTF-8");
+            }
+        }
+    }
+
+    /**
+     *
+     * Enter description here ...
+     * @param unknown_type $l_mid
```
Comment (sc979) [inline www/class/centreonGraphNg.class.php]: WIP

File: www/class/centreonGraphNg.class.php [hunk 9]
```diff
@@ -0,0 +1,1194 @@
...
+     */
+    public function getOVDColor($indexId, $metricId)
+    {
+        if (is_null($this->colorCache)) {
+            $this->colorCache = array();
+            
+            $stmt = $this->db->prepare("SELECT metric_id, rnd_color FROM `ods_view_details` WHERE `index_id` = :index_id");
```
Comment (sc979) [inline www/class/centreonGraphNg.class.php]: seems too long

File: www/class/centreonGraphNg.class.php [hunk 10]
```diff
@@ -0,0 +1,1194 @@
...
+            $stmt->bindParam(':index_id', $indexId, PDO::PARAM_INT);
+            $stmt->execute();
+            $this->colorCache = $stmt->fetchAll(PDO::FETCH_GROUP|PDO::FETCH_UNIQUE|PDO::FETCH_ASSOC);            
+        }
+        
+        if (isset($this->colorCache[$metricId]) && preg_match("/^\#[a-f0-9]{6,6}/i", $this->colorCache[$metricId]['rnd_color'])) {
```
Comment (sc979) [inline www/class/centreonGraphNg.class.php]: seems too long

File: www/class/centreonGraphNg.class.php [hunk 11]
```diff
@@ -0,0 +1,1194 @@
...
+            return $this->colorCache[$metricId]['rnd_color'];
+        }
+        $lRndcolor = $this->getRandomWebColor();
+        $stmt = $this->db->prepare("INSERT INTO `ods_view_details` (rnd_color, index_id, metric_id) VALUES (:rnd_color, :index_id, :metric_id)");
```
Comment (sc979) [inline www/class/centreonGraphNg.class.php]: seems too long

File: www/class/centreonGraphNg.class.php [hunk 12]
```diff
@@ -0,0 +1,1194 @@
...
+        $stmt->bindParam(':rnd_color', $lRndcolor, PDO::PARAM_STR);
+        $stmt->bindParam(':index_id', $indexId, PDO::PARAM_INT);
+        $stmt->bindParam(':metric_id', $metricId, PDO::PARAM_INT);
+        $stmt->execute();
+        return $l_rndcolor;
+    }
+
+    /**
+     *
+     * Enter description here ...
```
Comment (sc979) [inline www/class/centreonGraphNg.class.php]: WIP

File: www/class/centreonGraphNg.class.php [hunk 13]
```diff
@@ -0,0 +1,1194 @@
...
+            '#ff33ff', '#ff6600', '#ff6633', '#ff6666', '#ff6699', '#ff66cc',
+            '#ff66ff', '#ff9900', '#ff9933', '#ff9966', '#ff9999', '#ff99cc',
+            '#ff99ff', '#ffcc00', '#ffcc33', '#ffcc66', '#ffcc99', '#ffcccc',
+            '#ffccff');
+            return $webSafeColors[rand(0,sizeof($webSafeColors)-1)];
+    }
+
+    /**
+     *
+     * Enter description here ...
```
Comment (sc979) [inline www/class/centreonGraphNg.class.php]: WIP

File: www/class/centreonGraphNg.class.php [hunk 14]
```diff
@@ -0,0 +1,1194 @@
...
+                return 1;
+            }
+        }
+        return strnatcasecmp((isset($a["legend"]) && $a["legend"]) ? $a["legend"] : null,
+                             (isset($b["legend"]) && $b["legend"]) ? $b["legend"] : null);
+    }
+
+    /**
+     *
+     * Enter description here ...
```
Comment (sc979) [inline www/class/centreonGraphNg.class.php]: WIP

File: www/class/centreonGraphNg.class.php [hunk 15]
```diff
@@ -0,0 +1,1194 @@
...
+     * @param unknown_type $message
+     */
+    private function log($message)
+    {
+        if ($this->generalOpt['debug_rrdtool']['value'] && 
+            is_writable($this->generalOpt['debug_path']['value'])) {
+            error_log("[" . date("d/m/Y H:i") ."] RDDTOOL : ".$message." \n", 3,
+                      $this->generalOpt['debug_path']['value'] . "rrdtool.log");
```
Comment (sc979) [inline www/class/centreonGraphNg.class.php]: odd indentation

File: www/class/centreonGraphNg.class.php [hunk 16]
```diff
@@ -0,0 +1,1194 @@
...
+        }
+    }
+
+    /**
+     *
+     * Enter description here ...
```
Comment (sc979) [inline www/class/centreonGraphNg.class.php]: WIP

File: www/class/centreonGraphNg.class.php [hunk 17]
```diff
@@ -0,0 +1,1194 @@
...
+    /**
+     * Flush metrics in rrdcached
+     *
+     * @param array $metricsId The list of metrics
+     * @return bool
+     */
+    protected function flushRrdcached($metricsId)
+    {
+        if (!isset($this->generalOpt['rrdcached_enable']['value'])
+            || $this->generalOpt['rrdcached_enable']['value'] == 0) {
```
Comment (sc979) [inline www/class/centreonGraphNg.class.php]:
```suggestion
            || $this->generalOpt['rrdcached_enable']['value'] == 0
            ) {
```

File: www/class/centreonGraphNg.class.php [hunk 18]
```diff
@@ -0,0 +1,1194 @@
...
+            return true;
+        }
+
+        /*
+         * Connect to rrdcached
+         */
+        $errno = 0;
+        $errstr = '';
+        if (isset($this->generalOpt['rrdcached_port']['value'])
+            && trim($this->generalOpt['rrdcached_port']['value']) != '') {
```
Comment (sc979) [inline www/class/centreonGraphNg.class.php]:
```suggestion
            && trim($this->generalOpt['rrdcached_port']['value']) != ''
            ) {
```

File: www/class/centreonGraphNg.class.php [hunk 19]
```diff
@@ -0,0 +1,1194 @@
...
+            $sock = fsockopen('127.0.0.1', trim($this->generalOpt['rrdcached_port']['value']), $errno, $errstr);
+        } elseif (isset($this->generalOpt['rrdcached_unix_path']['value'])
+            && trim($this->generalOpt['rrdcached_unix_path']['value']) != '') {
```
Comment (sc979) [inline www/class/centreonGraphNg.class.php]:
```suggestion
            && trim($this->generalOpt['rrdcached_unix_path']['value']) != ''
            ) {
```

File: www/class/centreonGraphNg.class.php [hunk 20]
```diff
@@ -0,0 +1,1194 @@
...
+    /**
+     * Returns index data id
+     * 
+     * @param int $hostId
+     * @param int $serviceId
+     * @return int
+     */
+    public function getIndexDataId($hostId, $serviceId)
+    {
+        $stmt = $this->dbCs->prepare("SELECT id FROM index_data WHERE host_id = :host_id AND service_id = :service_id");
```
Comment (sc979) [inline www/class/centreonGraphNg.class.php]: seems too long

File: www/class/centreonGraphNg.class.php [hunk 21]
```diff
@@ -0,0 +1,1194 @@
...
+    private function addIndexId($indexId)
+    {
+        if (!isset($this->indexIds[$indexId])) {
+            $this->indexIds[$indexId] = 1;
+        }
+    }
+
+    private function getRealMetricsByIndexId($indexId)
+    {
+        $stmt = $this->dbCs->prepare("SELECT m.index_id, host_id, service_id, metric_id, metric_name, unit_name, min, max, warn, warn_low, crit, crit_low
```
Comment (sc979) [inline www/class/centreonGraphNg.class.php]: seems too long

File: www/class/centreonGraphNg.class.php [hunk 22]
```diff
@@ -0,0 +1,1194 @@
...
+                                       FROM metrics AS m, index_data AS i
+                                       WHERE i.id = :index_id
+                                       AND i.id = m.index_id 
+                                       AND m.hidden = '0'");
+        $stmt->bindParam(':index_id', $indexId, PDO::PARAM_INT);
+        $stmt->execute();
+        return $stmt->fetchAll(PDO::FETCH_ASSOC);
+    }
+    
+    private function getVirtualMetricsByIndexId($indexId)
```
Comment (sc979) [inline www/class/centreonGraphNg.class.php]: annotations

File: www/class/centreonGraphPoller.class.php [hunk 23]
```diff
@@ -124,22 +112,31 @@ class centreonGraphPoller
      * @param \CentreonDB $db
      * @param \CentreonDB $dbMonitoring
      * @param type $pollerId
```
Comment (sc979) [inline www/class/centreonGraphPoller.class.php]: deleted

File: www/class/centreonGraphPoller.class.php [hunk 24]
```diff
@@ -124,22 +112,31 @@ class centreonGraphPoller
...
-     * @param type $userId
      * @param type $start
```
Comment (sc979) [inline www/class/centreonGraphPoller.class.php]: deleted

File: www/class/centreonGraphPoller.class.php [hunk 25]
```diff
@@ -124,22 +112,31 @@ class centreonGraphPoller
...
      * @param type $end
```
Comment (sc979) [inline www/class/centreonGraphPoller.class.php]: deleted

File: www/class/centreonGraphPoller.class.php [hunk 26]
```diff
@@ -124,22 +112,31 @@ class centreonGraphPoller
...
-        $this->userId = $userId;
-        $this->start = $start;
-        $this->end = $end;
-
+                
         $this->initGraphOptions();
         $this->initRrd();
     }
+    
+    public function setPoller($pollerId, $graphName)
```
Comment (sc979) [inline www/class/centreonGraphPoller.class.php]: annotations

File: www/class/centreonGraphPoller.class.php [hunk 27]
```diff
@@ -240,21 +236,43 @@ public function setGraphName($graphName = '')
     {
         $this->graphName = $graphName;
     }
+    
+    public function addArgument($arg)
```
Comment (sc979) [inline www/class/centreonGraphPoller.class.php]: $arg type ?

File: www/class/centreonGraphPoller.class.php [hunk 28]
```diff
@@ -240,21 +236,43 @@ public function setGraphName($graphName = '')
...
+    {
+        $this->arguments[] = $arg;
+    }
+    
+    public function setRRDOption($name, $value = null)
```
Comment (sc979) [inline www/class/centreonGraphPoller.class.php]: annotations

File: www/class/centreonGraphPoller.class.php [hunk 29]
```diff
@@ -240,21 +236,43 @@ public function setGraphName($graphName = '')
...
+    {
+        if (strpos($value, " ")!==false) {
+            $value = "'".$value."'";
+        }
+        $this->rrdOptions[$name] = $value;
+    }
+    
+    private function log($message)
```
Comment (sc979) [inline www/class/centreonGraphPoller.class.php]: annotations

File: www/class/centreonGraphPoller.class.php [hunk 30]
```diff
@@ -240,21 +236,43 @@ public function setGraphName($graphName = '')
...
+        if ($this->generalOpt['debug_rrdtool'] && 
+            is_writable($this->generalOpt['debug_path'])) {
+            error_log("[" . date("d/m/Y H:i") ."] RDDTOOL : ".$message." \n", 3,
+                      $this->generalOpt['debug_path'] . "rrdtool.log");
+        }
+    }
 
     /**
      *
      * @param int $rows
```
Comment (sc979) [inline www/class/centreonGraphPoller.class.php]: deleted

File: www/class/centreonGraphPoller.class.php [hunk 31]
```diff
@@ -240,21 +236,43 @@ public function setGraphName($graphName = '')
...
      * @throws RuntimeException
      */
-    public function buildCommandLine($rows)
+    public function buildCommandLine($start, $end)
```
Comment (sc979) [inline www/class/centreonGraphPoller.class.php]: vars type

File: www/class/centreonGraphPoller.class.php [hunk 32]
```diff
@@ -310,80 +310,103 @@ public function buildCommandLine($rows)
      * @return array
      * @throws RuntimeException
      */
-    public function getData($rows = 200)
+    public function getGraph($start, $end)
```
Comment (sc979) [inline www/class/centreonGraphPoller.class.php]: vars type

File: www/class/centreonGraphPoller.class.php [hunk 33]
```diff
@@ -310,80 +310,103 @@ public function buildCommandLine($rows)
...
+    {
+        $this->buildCommandLine($start, $end);
+        return $this->getJsonStream();
+    }
+    
+    private function getJsonStream()
```
Comment (sc979) [inline www/class/centreonGraphPoller.class.php]: annotations

File: www/class/centreonGraphPoller.class.php [hunk 34]
```diff
@@ -310,80 +310,103 @@ public function buildCommandLine($rows)
...
+        if (is_writable($this->generalOpt['debug_path'])) {
+            $stderr = array('file', $this->generalOpt['debug_path'] . '/rrdtool.log', 'a');
+        } else {
+            $stderr = array('pipe', 'a');
+        }
         $descriptorspec = array(
-            0 => array('pipe', 'r'),
-            1 => array('pipe', 'w'),
-            2 => array('pipe', 'a'),
+                            0 => array("pipe", "r"),  // stdin est un pipe processus va lire
```
Comment (sc979) [inline www/class/centreonGraphPoller.class.php]:
```suggestion
                            0 => array("pipe", "r"),  // stdin - pipe process to read
```

File: www/class/centreonGraphPoller.class.php [hunk 35]
```diff
@@ -310,80 +310,103 @@ public function buildCommandLine($rows)
...
+                            1 => array("pipe", "w"),  // stdout est un pipe processus va ecrire
```
Comment (sc979) [inline www/class/centreonGraphPoller.class.php]:
```suggestion
                            1 => array("pipe", "w"),  // stdout - pipe process to write
```

File: www/class/centreonGraphPoller.class.php [hunk 36]
```diff
@@ -310,80 +310,103 @@ public function buildCommandLine($rows)
...
+                            2 => $stderr // stderr est un fichier
```
Comment (sc979) [inline www/class/centreonGraphPoller.class.php]:
```suggestion
                            2 => $stderr // stderr is a file
```

File: www/class/centreonGraphPoller.class.php [hunk 37]
```diff
@@ -310,80 +310,103 @@ public function buildCommandLine($rows)
...
-        /* Transform XML to values */
-        $xml = simplexml_load_string($str);
-        if (false === $xml) {
-            throw new RuntimeException();
+        
+        $this->formatByMetrics($rrdData);
+        return $this->graphData;
+    }
+    
+    private function formatByMetrics($rrdData)
```
Comment (sc979) [inline www/class/centreonGraphPoller.class.php]: annotations

**Comment by lpinsivy** [comment]
@garnier-quentin do you managed https://github.com/centreon/centreon/pull/5804?

**Comment by garnier-quentin** [comment]
Maybe. To be tested.

**Comment by garnier-quentin** [comment]
Done

**Comment by lpinsivy** [comment]
@garnier-quentin and also https://github.com/centreon/centreon/pull/6136?

File: www/Themes/Centreon-2/style.css [hunk 38]
```diff
@@ -2566,56 +2566,6 @@ ul.module_list {
     color: #88b917;
 }
 
-/* Tour plugin */
```
[the remark this answers is not in the document]
Comment (sc979) [reply www/Themes/Centreon-2/style.css]: ok : tourbus will be removed...

File: www/class/centreonGraph.class.php [hunk 39]
```diff
@@ -1493,36 +1494,23 @@ public function checkArgument($name, $tab, $defaultValue)
      * Enter description here ...
      * @param unknown_type $l_mid
```
Comment (sc979) [inline www/class/centreonGraph.class.php]:
```suggestion
     * @param sting $metricId
```

File: www/class/centreonGraphNg.class.php [hunk 40]
```diff
@@ -0,0 +1,1356 @@
...
+        );
+        
+        $this->cacheAllMetrics['r:' . $metric["metric_name"]] = $metric["metric_id"];
+
+        $dsData = $this->getCurveDsConfig($metric);
+        $this->metrics[$metric['metric_id']]['ds_data'] = $dsData;
+
+        $this->metrics[$metric['metric_id']]['legend'] = $this->getLegend($this->metrics[$metric["metric_id"]]);
+
+        $this->metrics[$metric['metric_id']]["stack"] = (isset($dsData["ds_stack"]) && $dsData["ds_stack"] ? $dsData["ds_stack"] : 0);
```
Comment (sc979) [inline www/class/centreonGraphNg.class.php]: line too long

File: www/class/centreonGraphNg.class.php [hunk 41]
```diff
@@ -0,0 +1,1356 @@
...
+        $this->metrics[$metric["metric_id"]]["crit"] = $metric["crit"];
+        $this->metrics[$metric["metric_id"]]["crit_low"] = $metric["crit_low"];
+        if (!isset($dsData["ds_color_area_warn"]) || empty($dsData["ds_color_area_warn"])) {
+            $this->metrics[$metric["metric_id"]]["ds_color_area_warn"] = $this->generalOpt["color_warning"]['value'];
+        }
+        if (!isset($dsData["ds_color_area_crit"]) || empty($dsData["ds_color_area_crit"])) {
+            $this->metrics[$metric["metric_id"]]["ds_color_area_crit"] = $this->generalOpt["color_critical"]['value'];
+        }
+        
+        $this->metrics[$metric["metric_id"]]["ds_order"] = (isset($dsData["ds_order"]) && $dsData["ds_order"] ? $dsData["ds_order"] : 0);
```
Comment (sc979) [inline www/class/centreonGraphNg.class.php]: line too long

File: www/class/centreonGraphPoller.class.php [hunk 42]
```diff
@@ -241,20 +246,67 @@ public function setGraphName($graphName = '')
...
+                "[" . date("d/m/Y H:i") ."] RDDTOOL : ".$message." \n",
+                3,
+                $this->generalOpt['debug_path'] . "rrdtool.log"
+            );
+        }
+    }
+
     /**
      *
      * @param int $rows
```
Comment (sc979) [inline www/class/centreonGraphPoller.class.php]:
```suggestion
      * @param unknow $start
      * @param unknow $end
      * @return void
```

File: www/class/centreonGraphPoller.class.php [hunk 43]
```diff
@@ -264,126 +316,146 @@ public function buildCommandLine($rows)
...
-            $this->metricsInfos[] = $info;
-
             $i++;
         }
     }
 
     /**
+     * Get graph result
      *
      * @param int $rows
```
Comment (sc979) [inline www/class/centreonGraphPoller.class.php]:
```suggestion
      * @param unknow $start
      * @param unknow $end
```

File: www/class/centreonGraphStatus.class.php [hunk 44]
```diff
@@ -59,134 +59,86 @@ public function __construct($index, $start, $end)
     /**
      * Get the metrics
      *
-     * @param integer $rows The number of points returned (Default: 200)
-     * @param array
+     * @param int   $rows The number of points returned (Default: 200)
```
Comment (sc979) [inline www/class/centreonGraphStatus.class.php]:
```suggestion
```

File: www/class/centreonGraphStatus.class.php [hunk 45]
```diff
@@ -59,134 +59,86 @@ public function __construct($index, $start, $end)
...
+     * @param mixed
```
Comment (sc979) [inline www/class/centreonGraphStatus.class.php]:
```suggestion
     * @return mixed $metrics
```

**Comment by ganoze** [comment]
@garnier-quentin Do these changes **require** rrdtool >= _1.5.x_ or is it **compatible** with it (and also with _1.4.x_).

**Comment by garnier-quentin** [comment]
It requires rrdtool >= 1.5.x (not compatible 1.4.x). I have provided the SPEC and sources to build it on centos 7. On debian stretch, the current version is ok (also in opensuse).

File: www/install/sql/centreon/Update-DB-18.10.4_to_19.04.0.sql [hunk 46]
```diff
@@ -1,2 +1,5 @@
 -- Change version of Centreon
 UPDATE `informations` SET `value` = '19.04.0' WHERE CONVERT( `informations`.`key` USING utf8 ) = 'version' AND CONVERT ( `informations`.`value` USING utf8 ) = '18.10.4' LIMIT 1;
+
+TRUNCATE TABLE ods_view_details;
+ALTER TABLE ods_view_details MODIFY metric_id int(11);
```
Comment (kduret) [inline www/install/sql/centreon/Update-DB-18.10.4_to_19.04.0.sql]: needs to be done in createTable.sql too for fresh install

**Comment by garnier-quentin** [comment]
I close that one. I have done one with a branch

**Pull request closed without merging**